### PR TITLE
Switch to (Re)TestItems

### PR DESCRIPTION
**Pull request by ToucheSir**
The impetus for this PR was twofold:
 1. Make use of the dedicated Julia test action as mentioned in https://github.com/FluxML/Metalhead.jl/pull/260#issuecomment-1841049671.
 2. Migrate off ReTest, which has not seen any activity in 2 years.

Along the way, I found some additional changes which could either be tackled here or in a follow-up PR:
 - We don't actually test any CUDA code on GPU CI! Buildkite merely runs the CPU test suite end-to-end, which feels wasteful
 - We don't have any tests which involve GPU arrays
 - We weren't testing `WideResNet` on GHA (fixed)
 - We aren't testing 1.6 outside of doctests

My feeling is that we'd want to set aside a subset of faster tests for 1.6/nightly/GPU CI. Maybe the smallest variant of each model. Then we can decrease our overall runtime while expanding our version matrix to cover everything we probably should've been covering.

### PR Checklist

- [x] Tests are added
- [x] Documentation, if applicable

### CODE REVIEW

File: test/Project.toml [hunk 1]
```diff
@@ -2,5 +2,16 @@
 CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
 Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
 Images = "916415d5-f1e6-5110-898d-aaa5f9f070e0"
-ReTest = "e0db7c4e-2690-44b9-bad6-7687da720f89"
+ReTestItems = "817f1d60-ba6b-4fd5-9520-3cf149f6a823"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+TestItems = "1c621080-faea-4a02-84b6-bbd5e436b8fe"
```
Comment (darsnack) [inline test/Project.toml:7]: My impression is that we don't need TestItems if we have ReTestItems?

Comment (ToucheSir) [reply test/Project.toml:7]: I figured that out after CI started complaining :)

The big problem with ReTestItems right now is that it doesn't support 1.6. My thought is to switch to TestItemRunner.jl, but that would mean we miss out on parallel execution (most importantly parallel _compilation_) for Buildkite. If https://github.com/JuliaTesting/ReTestItems.jl/issues/23 is addressed, perhaps we can revisit.

Comment (darsnack) [reply test/Project.toml:7]: We could use TestItemRunner for 1.6 and ReTestItems for the rest?

File: test/convnet_tests.jl [hunk 2]
```diff
@@ -316,7 +294,10 @@ end
             _gc()
         end
     end
-    @testset "MNASNet" verbose = true begin
+end
+
+@testitem "MNASNet" setup=[TestModels, TestMobileNets] begin
+    @testset for width in WIDTH_MULTS
```
Comment (darsnack) [inline test/convnet_tests.jl:300]:
```suggestion
    @testset for width_mult in WIDTH_MULTS
```

**Comment by ToucheSir** [comment]
Still needs a bit of work for GPU CI on 1.6 and nightly (possibly disabling the latter for now), but this is mostly good to go.

Some timings:
<table border="0" cellspacing="0" cellpadding="0" class="ta1"><colgroup><col width="324"/><col width="99"/><col width="387"/><col width="244"/></colgroup><tr class="ro1"><td style="text-align:left;width:7.415cm; " class="Default"><p>Group</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>Model</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>GHA time (PR)</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>Buildkite time (master)</p></td></tr><tr class="ro1"><td rowspan="2" style="text-align:left;width:7.415cm; " class="Default"><p>AlexNet|VGG</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>AlexNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>47.5s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>47.3s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>VGG</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>2m56.7s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>6m54.1s</p></td></tr><tr class="ro1"><td rowspan="4" style="text-align:left;width:7.415cm; " class="Default"><p>GoogLeNet|SqueezeNet|MobileNet|MNASNet</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>GoogLeNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>5m03.4s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>9m56.8s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>SqueezeNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>32.3s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>46.9s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>MobileNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>32.4s + 52.1s + 2m32.0s</p></td><td rowspan="2" style="text-align:left;width:5.59cm; " class="Default"><p>5m29.0s + 57.7s + 13.2s + 18.1s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>MNASNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>1m38.3s</p></td></tr><tr class="ro1"><td style="text-align:left;width:7.415cm; " class="Default"><p>EfficientNet</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>EfficientNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>8m00.6s + 6m46.4s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>6m51.6s + 4m42.0s</p></td></tr><tr class="ro1"><td rowspan="2" style="text-align:left;width:7.415cm; " class="Default"><p>ResNet|WideResNet</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>ResNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>16m36.8s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>16m41.3s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>WideResNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>1m49.0s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>2m41.8s</p></td></tr><tr class="ro1"><td style="text-align:left;width:7.415cm; " class="Default"><p>ResNeXt</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>ResNeXt</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>8m37.2s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>4m40.0s</p></td></tr><tr class="ro1"><td rowspan="2" style="text-align:left;width:7.415cm; " class="Default"><p>SEResNet|SEResNeXt</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>SEResNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>7m25.8s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>6m25.1s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>SEResNeXt</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>2m20.6s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>2m35.5s</p></td></tr><tr class="ro1"><td rowspan="2" style="text-align:left;width:7.415cm; " class="Default"><p>Res2Net|Res2NeXt</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>Res2Net</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>8m49.7s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>7m53.6s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>Res2NeXt</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>34.3s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>23.7s</p></td></tr><tr class="ro1"><td style="text-align:left;width:7.415cm; " class="Default"><p>Inception</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>Inception</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>7m31.9s = 2m12.7s + 1m34.7s + 2m26.1s + 1m10.6s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>6m52.9s</p></td></tr><tr class="ro1"><td style="text-align:left;width:7.415cm; " class="Default"><p>DenseNet</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>DenseNet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>8m56.4s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>8m48.7s</p></td></tr><tr class="ro1"><td style="text-align:left;width:7.415cm; " class="Default"><p>Unet</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>Unet</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>3m44.7s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>3m21.8s</p></td></tr><tr class="ro1"><td rowspan="2" style="text-align:left;width:7.415cm; " class="Default"><p>ConvNeXt|ConvMixer</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>ConvNeXt</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>4m40.1s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>6m06.1s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>ConvMixer</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>3m43.5s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>4m16.3s</p></td></tr><tr class="ro1"><td rowspan="3" style="text-align:left;width:7.415cm; " class="Default"><p>MLP-Mixer|ResMLP|gMLP</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>MLP-Mixer</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>3m10.9s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>3m39.8s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>ResMLP</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>1m58.4s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>2m31.7s</p></td></tr><tr class="ro1"><td style="text-align:left;width:2.258cm; " class="Default"><p>gMLP</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>1m58.9s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>2m19.5s</p></td></tr><tr class="ro1"><td style="text-align:left;width:7.415cm; " class="Default"><p>ViT</p></td><td style="text-align:left;width:2.258cm; " class="Default"><p>ViT</p></td><td style="text-align:left;width:8.858cm; " class="Default"><p>1m34.4s</p></td><td style="text-align:left;width:5.59cm; " class="Default"><p>1m02.0s</p></td></tr></table>

It appears we spend a lot of time compiling, as evidenced by the large time savings when similar models are run one after another. ViTs are an outlier despite their relative runtime slowness because they use the (type unstable under AD) Vector `Chain`. I wonder if we should explore expanding that to more models or having a serious look into other ideas for reducing compile times, but that's a discussion for another PR.

**Comment by theabhirath** [comment]
>It appears we spend a lot of time compiling, as evidenced by the large time savings when similar models are run one after another. ViTs are an outlier despite their relative runtime slowness because they use the (type unstable under AD) Vector Chain. I wonder if we should explore expanding that to more models or having a serious look into other ideas for reducing compile times, but that's a discussion for another PR.

During my GSoC, we explored this and I had noticed that when training, the Vector Chain gave me extremely bumpy loss curves – one of the reason we removed them from 0.7 to 0.8. A lot of this can come back slowly if we train more to isolate the exact problem, I think.

**Comment by ToucheSir** [comment]
With the renewed interest in https://github.com/FluxML/Metalhead.jl/pull/198#issuecomment-1846064419, now may be the time to revisit what's causing these mysterious instabilities during training. Shall we continue the discussion there?

File: test/model_tests.jl [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    (ViT, :base, (32, 32)),
+    (ViT, :large, (16, 16)),
+    (ViT, :large, (32, 32)),
+    (VGG, 11, false),
+    (VGG, 13, false),
+    (VGG, 16, false),
+    (VGG, 19, false),
+]
+
+const TEST_FAST = get(ENV, "FAST_TEST", "false") == "true"
```
Comment (darsnack) [inline test/model_tests.jl:45]:
```suggestion
const TEST_FAST = get(ENV, "TEST_FAST", "false") == "true"
```

Comment (ToucheSir) [reply test/model_tests.jl:45]: Woops, thanks!

**Comment by ToucheSir** [comment]
Ok, Buildkite is happy and so am I. This should be good to go. We now should have a pretty good picture of what works and doesn't on GPU too!

**Review by darsnack — APPROVED** [review]

**Pull request merged**